File: articles/differential-equations-for-systems-modeling/julia/logistic_ode.jl
# Differential Equations for Systems Modeling in Julia
# Logistic ODE simulation using Euler's method.
# Educational example only.

function logistic_rate(state, growth_rate, capacity)
    return growth_rate * state * (1 - state / capacity)
end

function simulate_logistic(initial_state, growth_rate, capacity, dt, steps)
    time = zeros(steps)
    state = zeros(steps)

    state[1] = initial_state

    for i in 2:steps
        derivative = logistic_rate(state[i - 1], growth_rate, capacity)
        state[i] = state[i - 1] + derivative * dt
        time[i] = time[i - 1] + dt
    end

    return time, state
end

time, state = simulate_logistic(10.0, 0.20, 100.0, 0.1, 300)

println("Final state: ", state[end])
println("Maximum state: ", maximum(state))

File: articles/differential-equations-for-systems-modeling/fortran/logistic_ode.f90
program logistic_ode
  implicit none

  integer, parameter :: steps = 300
  integer :: i
  real :: dt, growth_rate, capacity, derivative
  real, dimension(steps) :: state

  dt = 0.1
  growth_rate = 0.20
  capacity = 100.0
  state(1) = 10.0

  do i = 2, steps
    derivative = growth_rate * state(i - 1) * (1.0 - state(i - 1) / capacity)
    state(i) = state(i - 1) + derivative * dt
  end do

  print *, "Final state:", state(steps)

end program logistic_ode
